Sources/Apollo/HTTPNetworkTransport.swift

```diff
@@ -1,11 +1,32 @@
 import Foundation
 
+public protocol HTTPNetworkTransportDelegate: class {
```

Comment:
I've blocked this way too long and I want to leave it to you to decide what to do, but [in the past](https://github.com/apollographql/apollo-ios/pull/297#issuecomment-400535981) I've seen it as a hard requirement that our solution should also support retries on error and token refresh, which probably requires at least some callbacks to be asynchronous. See [here](https://github.com/Alamofire/Alamofire/blob/master/Documentation/AdvancedUsage.md#adapting-and-retrying-requests) for the way Alamofire supports this for example. Maybe it would be enough to make `shouldSend` asynchronous?

Reply:
Here are some other previous ideas for delegate methods: https://github.com/apollographql/apollo-ios/pull/257#issuecomment-381837750

Reply:
There was also this idea about propagating custom attributes from client API methods in a context argument: https://github.com/apollographql/apollo-ios/pull/257#issuecomment-381400610

Reply:
Not saying you should do anything with this, I know I've probably made it too complicated in the past :)

Reply:
retries on token error and refresh is supper useful, we have that in our code. Helps with wrong dates on client and token format changes on server

Reply:
I based a bunch of this on [your suggestions here](https://github.com/apollographql/apollo-ios/pull/257#issuecomment-381837750). 

What am I missing about `Context` that wouldn't be covered by just altering the `Request` object? 

I'm gonna take a look at making response handling a separate delegate, since not everyone who needs request handling needs response handling, and optional delegate methods require all sorts of objc nonsense. 

I'll admit that I'm kiiiind of reluctant to do custom retry logic due to the additional complexity, but I'll see if I can make that work.

Reply:
Response handling and retry delegates have now been added.

Reply:
Oh, I missed this before, but to answer your question about `Context`: this would allow you to pass in additional data from `ApolloClient` methods like `fetch` (so per request, at the point of use), as opposed to doing it from the delegate. Probably not needed right now, but for context :)